### ISSUE TITLE
SQL query editor: Special characters in column names not supported

### ISSUE THREAD

**Issue by markus-ksg** [phpmyadmin/phpmyadmin#16314]
### Describe the bug

The Ctrl+Space code completion in the SQL query editor doesn't support special characters in column names very well.

### To Reproduce

1. Go to https://demo.phpmyadmin.net/master-config/index.php?route=/table/sql&db=burger_code&table=categories&server=1
2. Add columns `ITM-main#` and `#main`  (names contain special characters):
```sql
ALTER TABLE `categories` ADD `ITM-main#` INT NOT NULL AFTER `name`; 
ALTER TABLE `categories` ADD `#main` INT NOT NULL AFTER `name`; 
```
3. Try to use the columns in SQL query completion (Ctrl+Space). Examples after inital query text ( `|` is the prefix/completion position; press Ctrl+Space, choose ITM-main#  or #main from the popup):
```sql
SELECT * FROM `categories` WHERE |
```
prefix       | result | expected
----------|-------|----------
(none)     | `` ITM-main# `` | `` `ITM-main#` ``
`IT`          | `` ITM-main# `` | `` `ITM-main#` ``
`ITM-`     | `` ITM-ITM-main# `` | `` `ITM-main#` `` (or `` ITM-`ITM-main#` `` if you consider the `-` a difference operator)
`` ` ``        | `` `ITM-main#` `` | `` `ITM-main#` `` (good)
`` `ITM ``  | `` `ITM-main#` `` | `` `ITM-main#` `` (good)
`` `ITM- ``  | `` `ITM-ITM-main# `` | `` `ITM-main#` ``
`` `ITM-ma ``  | no useful proposals | `` `ITM-main#` ``
`` # ``       | `` ##main `` | `` `#main` ``
`` #ma ``  | `` #ma#main `` | `` `#main` ``
`` ` ``        | `` `#main` `` | `` `#main` `` (good)
`` `# ``      | `` `##main `` |  `` `#main` ``
`` `#ma `` | `` `#ma#main `` |  `` `#main` ``

### Expected behavior

Completion of special column names should work like for columns with regular names. After choosing something from the popup, the column name should be completed and enclosed in `` ` ` `` if necessary.

### Server configuration

- phpMyAdmin version:  5.0.2, 5.1.0-dev, and older versions

**Comment by markus-ksg** [comment]
To reproduce if the example database `burger_code` doesn't exist, just create the database and then create table `categories` with column `name`.

**Comment by williamdes** [comment]
Thank you for reporting this issue, I hope someone will have a look into it :)

**Comment by shucon** [comment]
Starting to work on this.

**Comment by williamdes** [comment]
Hi @markus-ksg 

Would you agree to test the latest version in development [(phpMyAdmin 5.1+snapshot)](https://www.phpmyadmin.net/downloads/#devel) ?
I think this is fixed now

**Comment by markus-ksg** [comment]
I'd love to test it, but I couldn't find a Docker image that includes the latest 5.1 code.

I'm currently using the `phpmyadmin/phpmyadmin:latest` image, which contains 5.0.4. I tried to replace /var/www/html in the running docker container with the contents of https://files.phpmyadmin.net/snapshots/phpMyAdmin-5.1+snapshot-english.tar.gz . After `chown -R --reference=html.bak html ; chmod --reference=html.bak html`, I did get the login page, but after entering the credentials, I got redirected to /index.php?route=/ with error message "mysqli::real_connect(): (HY000/2002): No such file or directory".
Do you have any hints on how I could test this locally?

https://demo.phpmyadmin.net/QA_5_1 shows some general error messages. After ignoring them, the bug still reproduces.
https://demo.phpmyadmin.net/master/ (5.2) also still exhibits the described bad behavior.

I'm afraid https://github.com/codemirror/CodeMirror/issues/6331 was not enough to fix this.

**Comment by ibennetch** [comment]
> I'd love to test it, but I couldn't find a Docker image that includes the latest 5.1 code.

That's correct, we don't publish Docker images for prerelease or release candidate versions.

> chown -R --reference=html.bak html

You taught me a new argument to use with chown, nice!

> Do you have any hints on how I could test this locally?

Probably you need to change one line, in `libraries/vendor_config.php` find the line `define('CONFIG_DIR', ROOT_PATH);` and change it to `define('CONFIG_DIR', '/etc/phpmyadmin/');` (this is how the Docker image knows where to find the configuration files).

> https://demo.phpmyadmin.net/QA_5_1 shows some general error messages.

This should be cleared up now.

**Comment by markus-ksg** [comment]
> > Do you have any hints on how I could test this locally?
> 
> Probably you need to change one line, in `libraries/vendor_config.php` find the line `define('CONFIG_DIR', ROOT_PATH);` and change it to `define('CONFIG_DIR', '/etc/phpmyadmin/');` (this is how the Docker image knows where to find the configuration files).

Thanks, that was it! Great that I can now test pre-release versions in my Docker setup. 5.1 looks good so far, but as expected, this bug also shows up locally.
I guess it's a CodeMirror issue in the completion implementation, since there are cases where useful column proposals are computed, but insertion fails to find the correct starting position.

**Comment by ben-wckdrzr** [comment]
I found this page while looking for a solution to the same issue directly in CodeMirror - this is an issue in that package as mentioned, but doesn't seem to have been raised there.

I asked a question about this on the CodeMirror discussion (https://discuss.codemirror.net/t/quoting-table-names-when-using-schemacompletionsource/6721), and a fix was implemented in @codemirror/lang-sql v6.5.1

Not all of the test cases above pass, but most do; and enough for my needs

Just posting here to try and keep the internet tidy :D

**Comment by williamdes** [comment]
> Just posting here to try and keep the internet tidy :D

Thanks

I think this is related to #18521
